Observe layout guides

frederoni opened the pull request:
Fixes #6954

This PR adds observers to the top and bottom layout guides if necessary in order to re-layout after changes have occurred. (i.e due to showing/hiding a navigation bar)

It also replaces constraints with manual layout to avoid manipulating the view controller's view which is the developer's responsibility.

![animation](https://cloud.githubusercontent.com/assets/764476/22205062/7a4a45b6-e175-11e6-8053-5da3d03ae8d5.gif)

frederoni commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -1866,6 +1890,10 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
             }
         }
     }
+    else if ([keyPath isEqualToString:@"bounds"] && [object isKindOfClass:NSClassFromString(@"_UILayoutGuide")])
```
Comparing against the private class _UILayoutGuide is not particularly future proof.

1ec5 commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -122,6 +122,9 @@ typedef NS_ENUM(NSUInteger, MGLUserTrackingState) {
 const CGSize MGLAnnotationAccessibilityElementMinimumSize = CGSizeMake(10, 10);
 
+// Context for KVO observing UILayoutGuides.
+static char MGLLayoutGuidesUpdatedContext = 1;
```
This works, but the more common practice is to pass in a selector such as `_cmd`.

1ec5 replied:
Er, I may be mistaken about `_cmd`. But in any case, the important thing is to ensure that the context’s _value_ is unique. If we later add a key-value observer with a different constant set to 1, suddenly there’s an opportunity for crosstalk.

A simple fix is to [set the context equal to its own address](http://nshipster.com/key-value-observing/#correct-context-declarations).

1ec5 commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -655,6 +658,11 @@ - (void)dealloc
     [_attributionButton removeObserver:self forKeyPath:@"hidden"];
 
+    if (_isObservingLayoutGuides) {
+        [(UIView *)self.viewControllerForLayoutGuides.topLayoutGuide removeObserver:self forKeyPath:@"bounds" context:(void*)&MGLLayoutGuidesUpdatedContext];
```
Is topLayoutGuide really a UIView? I'd expect it to be a UILayoutGuide.

1ec5 commented:
Postponing to v3.4.1, per https://github.com/mapbox/mapbox-gl-native/issues/6954#issuecomment-273283732.

1ec5 commented:
Since this PR is scheduled for iOS SDK v3.5.0, it needs to be rebased and retargeted at the release-ios-v3.5.0-android-v5.0.0 branch.

frederoni commented:
- Rebased and retargeted against the 3.6 release branch.
- Fixed layout for the recently added scale bar.

![observe](https://cloud.githubusercontent.com/assets/764476/26172942/665de1aa-3b4a-11e7-8552-bc9173d86ac6.gif)

@boundsj @fabian-guerra 👀

1ec5 commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -931,6 +856,38 @@ - (void)layoutSubviews
 
+- (void)layoutOrnaments
+{
+    CGFloat x, y;
```
This is OK, but just a stylistic tip: `x` and `y` variables that get repurposed multiple times carry the risk that we might fail to update one of these variables and its previous value may “bleed” into a subsequent use of it.

Instead, I sometimes take advantage of the struct initializer syntax to keep the values scoped to where they’re used. For example:

```objc
self.compassView.center = {
    CGRectGetWidth(self.bounds) - CGRectGetMidX(self.compassView.bounds) - self.contentInset.right - 5,
    CGRectGetMidY(self.compassView.bounds) + self.contentInset.top + 5,
};
```

If you’re concerned about accidentally swapping `x` and `y`, you can also use the C99 designated initializer syntax:

```objc
self.compassView.center = {
    .x = CGRectGetWidth(self.bounds) - CGRectGetMidX(self.compassView.bounds) - self.contentInset.right - 5,
    .y = CGRectGetMidY(self.compassView.bounds) + self.contentInset.top + 5,
};
```
